mlflow/R/mlflow/R/tracking-runs.R

```diff
@@ -0,0 +1,597 @@
+                             source_name = source_name, entry_point_name = entry_point_name, start_time = start_time,
+                             source_version = source_version, tags = tags, experiment_id = experiment_id)
+    return(run)
+  }
+
+  # Fluent mode, check to see if extraneous params passed.
+
+  if (!is.null(user_id)) stop("`user_id` should only be specified when `client` is specified.", call. = FALSE)
+  if (!is.null(run_name)) stop("`run_name` should only be specified when `client` is specified.", call. = FALSE)
+  if (!is.null(start_time)) stop("`start_time` should only be specified when `client` is specified.", call. = FALSE)
+  if (!is.null(tags)) stop("`tags` should only be specified when `client` is specified.", call. = FALSE)
+
+  source_type <- source_type %||% "LOCAL"
+  active_run_id <- mlflow_get_active_run_id()
+  if (!is.null(active_run_id)) {
+    stop("Run with UUID ", active_run_id, " is already active.",
+         call. = FALSE
+    )
+  }
+
+  existing_run_uuid <- run_uuid %||% {
+    env_run_id <- Sys.getenv("MLFLOW_RUN_ID")
+    if (nchar(env_run_id)) env_run_id
+  }
+
+  client <- mlflow_client()
+
+  run <- if (!is.null(existing_run_uuid)) {
+    # This is meant to pick up existing run when we're inside `mlflow_source()` called via `mlflow run`.
+    mlflow_get_run(client = client, run_id = existing_run_uuid)
+  } else {
+    experiment_id <- mlflow_infer_experiment_id(experiment_id)
+    client <- mlflow_client()
+
+    args <- mlflow_get_run_context(
+      client,
+      experiment_id = experiment_id,
+      source_name = source_name,
+      source_version = source_version,
+      entry_point_name = entry_point_name,
+      source_type = source_type
+    )
+    do.call(mlflow_create_run, args)
+  }
+  mlflow_set_active_run_id(mlflow_id(run))
+  run
+}
+
+
+mlflow_get_run_context <- function(client, ...) {
+  UseMethod("mlflow_get_run_context")
+}
+
+mlflow_get_run_context.default <- function(client, source_name, source_version, experiment_id,
+                                           ...) {
+  list(
+    client = client,
+    source_name = source_name %||% get_source_name(),
+    source_version = source_version %||% get_source_version(),
+    experiment_id = experiment_id %||% 0,
+    ...
+  )
+}
+
+#' End a Run
+#'
+#' Terminates a run. Attempts to end the current active run if `run_id` is not specified.
+#'
+#' @param status Updated status of the run. Defaults to `FINISHED`.
+#' @param end_time Unix timestamp of when the run ended in milliseconds.
+#' @template roxlate-run-id
+#' @template roxlate-client
+#'
+#' @export
+mlflow_end_run <- function(status = c("FINISHED", "SCHEDULED", "FAILED", "KILLED"),
+                           end_time = NULL, run_id = NULL, client = NULL) {
+
+  status <- match.arg(status)
+  end_time <- end_time %||% current_time()
+
+  active_run_id <- mlflow_get_active_run_id()
+
+  if (!is.null(client) && is.null(run_id))
```

Comment:
should we just call resolve_client_and_run_id?

Reply:
`resolve_client_and_run_id()` could start a run though